spec.md

```diff
@@ -778,6 +800,182 @@ message Volume {
+  // from the list of permitted topologies.
+  // If the list of permitted topologies is not specified, then the SP
+  // may choose from the list of all possible topologies.
+  // If the list of permitted topologies is specified and the SP is
+  // unable to to make the provisioned volume available from any of the
+  // preferred topologies it MUST fail the CreateVolume call.
+  // 
+  // Example 1:
+  // Given a volume should be accessible from a single zone, and
+  // permitted =
+  //   {"region": "R1", "zone": "Z2"},
+  //   {"region": "R1", "zone": "Z3"}
+  // preferred =
+  //   {"region": "R1", "zone": "Z3"}
+  // then the the SP SHOULD first attempt to make the provisioned volume
+  // available from "zone" "Z3" in the "region" "R1" and fall back to
+  // "zone" "Z2" in the "region" "R1" if that is not possible.
+  //
+  // Example 2:
+  // Given a volume should be accessible from a single zone, and
+  // permitted =
+  //   {"region": "R1", "zone": "Z2"},
+  //   {"region": "R1", "zone": "Z3"},
+  //   {"region": "R1", "zone": "Z4"},
+  //   {"region": "R1", "zone": "Z5"}
+  // preferred =
+  //   {"region": "R1", "zone": "Z4"},
+  //   {"region": "R1", "zone": "Z2"}
+  // then the the SP SHOULD first attempt to make the provisioned volume
+  // accessible from "zone" "Z4" in the "region" "R1" and fall back to
+  // "zone" "Z2" in the "region" "R1" if that is not possible. If that
+  // is not possible, the SP may choose between either the "zone"
+  // "Z3" or "Z5" in the "region" "R1".
+  //
+  // Example 3:
+  // Given a volume should be accessible from TWO zones, and
+  // permitted =
+  //   {"region": "R1", "zone": "Z2"},
+  //   {"region": "R1", "zone": "Z3"},
+  //   {"region": "R1", "zone": "Z4"},
+  //   {"region": "R1", "zone": "Z5"}
+  // preferred =
+  //   {"region": "R1", "zone": "Z5"},
+  //   {"region": "R1", "zone": "Z3"}
+  // then the the SP SHOULD first attempt to make the provisioned volume
+  // accessible from the combination of the two "zones" "Z5" and "Z3" in
+  // the "region" "R1". If that's not possible, it should fall back to
+  // a combination of "Z5" and other possibilities from the list of
+  // permitted. If that's not possible, it should fall back  to a
+  // combination of "Z3" and other possibilities from the list of
+  // permitted. If that's not possible, it should fall back  to a
+  // combination of other possibilities from the list of permitted.
+  repeated Topology preferred = 2;
+}
+
+// Topology is a map of topological domains to topological segments.
+// A topological domain is a sub-division of a cluster, like "region",
+// "zone", "rack", etc.
+// A topological segment is a specific instance of a topological domain,
+// like "zone3", "rack3", etc.
+// For example {"zone": "Z1", "rack": "R3"}
+// Each key (topological domain) must be 63 characters or less and
+// consist of alphanumeric characters, '-', '_' or '.'.
+// Each key SHOULD include the plugin's host company name and the plugin
```

Comment:
There's a bit of redundancy here w/ respect to the plugin name constraints than we've already documented in the `GetPluginInfoResponse` message, along with an implicit tightening of said constraints: if a key is to include a combination of host company + plugin name AND the topo domain key name AND STILL be 63 chars or less, then `GetPluginInfoResponse.name` MUST be strictly less than 63 characters. This sentence pushes the boundaries w/ respect to violating the CSI v0.2 spec.
Furthermore, these keys are **opaque** to the CO. This sentence seems hand-crafted to suit the needs of particular CO implementation strategy, something that I'd rather avoid in the spec if possible. If a CO wants to convert these topo domain keys into components of an object label, then the CO should be responsible for avoiding collision between plugin namespaces - not the SP.
Instead of **implicitly** breaking the CSI v0.2 spec w/ key content+length requirements here, maybe we should *explicitly* break compatibility by changing the length constraints for plugin names to 32 chars or less, and specify here that the topo domain key names must be to 32 chars or less .. then if a CO wants to concatenate these bits, it can do so w/o worry.

Reply:
Even by adding org name, i'm not sure how it solves the "implementation leak" and implicit constraints that I'm complaining about here: the limits of the proposed implementation are still leaking into length constraints for fields in the spec. What's the length limit for an org name property? Such a limit would directly influence the length limit of the topo keys here, but only considering a particular implementation strategy. I'd like to avoid such implicit dependencies in the spec.
Instead of a 1st class org name field, maybe simply rewrite the constraints here such that topo keys are composed of an org identifier and the topo domain name. For example, `gce.region`. The length limit would be defined here, and here only - no other fields in the spec have an implicit effect on the length limitations defined here.

Reply:
Ok, so I will leave the wording about company name in the key as is, which would allow `com.google.region`. But will loosen the requirements to align with those of the plugin name? Does that sound ok?

Reply:
One suggestion: can we just say the keys here need to follow java package naming conventions (https://en.wikipedia.org/wiki/Java_package#Package_naming_conventions) to avoid potential conflicts? I thing the goals of these are the same, although I think `-` is disallowed in java package naming. Maybe we can just copy some of the spec text from there?